snakemake-slurm.md

```diff
@@ -0,0 +1,100 @@
+```
+Notes: 
+  - Most clusters would prefer if you run use an interactive session to run this, so that you're not running anything on the login nodes. Since this process is only submitting jobs, you _can_ run this command on tmux/screen on a login node, but only do it for a small number of jobs (otherwise you'll slow everyone up and your job will be killed by admin!).
+ - the `--jobs` parameter allows snakemake to submit up to NUM_JOBS_TO_SUBMIT number of jobs, but please be aware of submission limits on your cluster. By default, snakemake will only submit jobs that can be run (input files already exist). There is a parameter called `--immediate-submit` that will submit all jobs at once, but this may be an issue if the input files for those jobs are not available when those jobs make it through the scheduling queue.
+
+## Using a cluster configuration file
+
+To save time you can also make a yml file containing your sbatch information, and tell snakemake where to find it.
+
+Here's an example cluster configuration file:
+```
+# cluster_config.yml - cluster configuration
+__default__:
+    account: ACCOUNT
+    partition: PARTITION
+    time: 01:00:00 # time limit for each job
+    nodes: 1
+    ntasks-per-node: 14 #Request n cores be allocated per node.
+    chdir: /directory/to/change/to
+    output: a_name_for_my_job-%j.out
+    error: a_name_for_my_job-%j.err
+```
+
+Even if you tell snakemake where to find this file, it's not going to use all of these parameters to submit each job - it will only use the onse you specify in the `sbatch` portion of your `--cluster` statement.
+
+To submit with identical parameters as we used above, run snakemake like so:
+
+```
+snakemake --cluster "sbatch -A {cluster.account} -t {cluster.time} -p {cluster.partition} -N {cluster.nodes}" --cluster-config cluster_config.yml —jobs NUM_JOBS_TO_SUBMIT
+```
+The information within the `{}` are the parameters that snakemake will read from the `cluster_config.yml` file.
+
+In this configuration file above, I only have information for a `__default__`, which will be used as the default for each rule. If you want to set specific time limits for each rule (or some rules), you can add that info to the file. 
+
+For example, if I have a rule called `trimmomatic_raw`, I could add the following to my `cluster_config.yml` file to specify some different cluster parameters for that rule.
+
+```
+trimmomatic_raw:
+   time: 00:45:00 # time limit for this rule only
+```
+
+For non-slurm clusters, you can change the `cluster` command to reflect the scheduling service your cluster uses. See snakemake's documentation for examples.
+
+
+## Examples
+
+### Example to run within tmux
+```
+source ~/.bashrc
+conda activate snakemake
+
+cd /home/ntpierce/2019-burgers-shrooms/orthofinder_work
+
+snakemake -s diamond_blast.snakefile --use-conda --cluster "sbatch -t 0:30:00 -N 1 -c 14 -J dmnd --mem=30gb " --jobs 200
```

Comment:
Maybe don't show an example where you've asked for 200X30GB ram at once? That's 6TB, and our group limit is 1TB. You haven't specific a partition here so you'd be using any free space on the cluster, but still...that's a lot of resources.

Reply:
heh, yeah that was leftover from the other cluster. will fix, thx!